### PR TITLE
Add shipping partner suggestions api

### LINKED ISSUE #36281: Add remote endpoint API support for shipping partners in shipping task
This is a follow-up to https://github.com/woocommerce/woocommerce/issues/36248.

The objective of this task is to add a remote API endpoint for displaying shipping partners in the marketing task. This is to be hosted on WooCommerce.com, similar to `obw-free-extensions`. The data hosted are as follows:

1. Shipping partner logo
1. Plugin name
1. Plugin slug
1. Countries supported
1. Learn more link
1. Other copies depending on the confirmed design

We'd also need the image to be hosted on WooCommerce.com

## Acceptance Criteria

1. Add the ability for a remote endpoint to enable or disable the featured partners (p73Jol-1C5-p2#comment-6419)

### PR DESCRIPTION
**Pull request by moon0326**
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #36281

This PR adds `/wp-json/wp-admin/shipping-partner-suggestions`. The endpoint is designed to be used with the shipping cost task.

**Default shipping list test**

When `woocommerce_show_marketplace_suggestions` option is set to `no`, the endpoint should use the [default shipping partner list](https://github.com/woocommerce/woocommerce/blob/9d24bf5b37b7b3230a8b2672e69f3a2642e9f750/plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php)

1. Make sure `_transient_woocommerce_admin_shipping_partner_suggestions_specs` optoin does not exist.
2. Set `woocommerce_show_marketplace_suggestions` option to `no`
3. Acces `/wp-json/wp-admin/shipping-partner-suggestions`
4. Confirm the endpoint returns a valid JSON.
5. Confirm `_transient_woocommerce_admin_shipping_partner_suggestions_specs` option still does not exist.

**Getting the list from WCCOM**

1. Setup your local WCCOM env with https://github.com/Automattic/woocommerce.com/pull/16291 branch.
2. Open this [file](https://github.com/woocommerce/woocommerce/blob/77f78f8c9b938ca63b84b14a25db9c027aad69c5/plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/ShippingPartnerSuggestionsDataSourcePoller.php) and change the DATA_SOURCES to your local WCCOM.

3. Make sure `woocommerce_show_marketplace_suggestions` is set to `yes`
4. Make sure `_transient_woocommerce_admin_shipping_partner_suggestions_specs` does not exist. 
5. Acces `/wp-json/wp-admin/shipping-partner-suggestions`
6. Confirm the endpoint returns a valid JSON
7. Confirm `_transient_woocommerce_admin_shipping_partner_suggestions_specs` exists. 

*Depending on your local setup, you might need to set [WP_PROXY_HOST](https://developer.wordpress.org/reference/classes/wp_http_proxy/) to access your local WCCOM env from your WooCommerce dev.

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+				'layout_row'        => array(
+					'image'    => $asset_base_url . 'skydropx-row.svg',
+					'features' => $row_layout_features,
+				),
+				'description'       => '',
+				'learn_more_link'   => 'https://wordpress.org/plugins/skydropx-cotizador-y-envios/',
+				'is_visible'        => array(
+					self::get_rules_for_countries( array( 'MX', 'CO' ) ),
+				),
+				'available_layouts' => array( 'single' ),
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php]: Should it be "row" layout?

File: plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/ShippingPartnerSuggestions.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Class ShippingPartnerSuggestions
+ */
+class ShippingPartnerSuggestions {
+
+	/**
+	 * Go through the specs and run them.
+	 *
+	 * @param array|null $specs payment suggestion spec array.
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/ShippingPartnerSuggestions.php]:
```suggestion
	 * @param array|null $specs shipping partner suggestion spec array.
```

File: plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	 * @return \WP_Error|\WP_HTTP_Response|\WP_REST_Response
+	 */
+	public function get_suggestions( $request ) {
+		$should_display = $this->should_display();
+		$force_default  = $request->get_param( 'force_default_suggestions' );
+
+		if ( $should_display ) {
+			return Suggestions::get_suggestions();
+		} elseif ( false === $should_display && true === $force_default ) {
+			return rest_ensure_response( Suggestions::get_suggestions( DefaultPaymentGateways::get_all() ) );
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php]:
```suggestion
			return rest_ensure_response( Suggestions::get_suggestions( DefaultShippingPartners::get_all() ) );
```

File: plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => \WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_suggestions' ),
+					'permission_callback' => array( $this, 'get_permission_check' ),
+					'args'                => array(
+						'force_default_suggestions' => array(
+							'type'        => 'boolean',
+							'description' => __( 'Return the default shipping partner suggestions when woocommerce_show_shipping_partner_suggestions and woocommerce_setting_payments_recommendations_hidden options are set to no', 'woocommerce' ),
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php]: Do you know where we set `woocommerce_show_shipping_partner_suggestions` options? And I guess we don't need `woocommerce_setting_payments_recommendations_hidden`?

**Review by chihsuan — CHANGES_REQUESTED** [review]
Thanks for working on it! @moon0326 

I just left some comments and I think we should replace `DefaultPaymentGateways` with `DefaultShippingPartners`.

File: plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php (hunk 1 above)
Comment (moon0326) [reply plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php]: @chihsuan Thank you for catching it 🙏

File: plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php (hunk 4 above)
Comment (moon0326) [reply plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php]: That description needs an update 😄 It only works when `woocommerce_show_marketplace_suggestions` is set to false and `force_default_suggestions` is set to true 

Code is here: https://github.com/woocommerce/woocommerce/pull/37155/files#diff-d0513d5e143fc2d4484715f024daa474107356bc224e4fa6e20bea70c04c03fdR97

Comment (moon0326) [reply plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php]: @chihsuan Updated in https://github.com/woocommerce/woocommerce/pull/37155/commits/66255379865608ba47583d47f3c41d032fefa83b 🙏

File: plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php (hunk 1 above)
Comment (moon0326) [reply plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php]: Updated in https://github.com/woocommerce/woocommerce/pull/37155/commits/302ffe0e2dbb79dd905e5cf3956a253d8afa4275

File: plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+					'type'        => 'boolean',
+					'context'     => array( 'view', 'edit' ),
+					'readonly'    => true,
+				),
+				'available_layouts' => array(
+					'description' => __( 'Available layouts -- single, dual, or both', 'woocommerce' ),
+					'type'        => 'array',
+					'items'       => array(
+						'type' => 'string',
+						'enum' => array( 'row', 'column' ),
```
Comment (moon0326) [inline plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php:195]: FYI, I borrowed `row` and `column` concept from CSS `flex-direction`. 

Layout `row` represents a smaller banner (?) that can be placed with other smaller banners in a single row.
Layout `column` represents a banner that takes the whole row width.

File: plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+				),
+				'learn_more_link'   => array(
+					'description' => __( 'Learn more link .', 'woocommerce' ),
+					'type'        => 'string',
+					'context'     => array( 'view', 'edit' ),
+					'readonly'    => true,
+				),
+				'is_visible'        => array(
+					'description' => __( 'Suggestion visibility.', 'woocommerce' ),
+					'type'        => 'boolean',
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php:186]: @moon0326 I just noticed that `is_visible` is not a boolean value in the API response. Should we change it to boolean?

Screenshot 2023-03-14 at 09 53 21

**Comment by chihsuan** [comment]
BTW, looks like PHP linting is failing (also WCCOM side).

File: plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php (hunk 6 above)
Comment (moon0326) [reply plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php:186]: Good catch!

It should be set to boolean after rule evaluation. Updated in https://github.com/woocommerce/woocommerce/pull/37155/commits/88bf0713fd7c2a054592f1d02292204aa5384c7d

**Review by chihsuan — DISMISSED** [review]
Nice work! Looks good to me and tested well. 💯

File: plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php (hunk 5 above)
Comment (rjchow) [reply plugins/woocommerce/src/Admin/API/ShippingPartnerSuggestions.php:195]: I think i'd go with 'narrow' and 'wide' but no objections to this way either 😀

File: plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+ */
+class DefaultShippingPartners {
+
+	/**
+	 * Get default specs.
+	 *
+	 * @return array Default specs.
+	 */
+	public static function get_all() {
+		$asset_base_url      = WC()->plugin_url() . 'assets/images/shipping_partners/';
```
Comment (rjchow) [inline plugins/woocommerce/src/Admin/Features/ShippingPartnerSuggestions/DefaultShippingPartners.php]: I think there's a missing slash between `woocommerce` and `assets` in the path - I get a 404 with the URL that's generated but its found if i add the slash
image

**Review by rjchow — DISMISSED** [review]
Left two comments but lgtm overall, thanks! We can iterate on this with the UI work next 😄

**Review by chihsuan — APPROVED** [review]
Looks good!

**Pull request merged**